src/shogun/mathematics/linalg/Vector.cpp

```diff
@@ -0,0 +1,226 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/Vector.h>
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+template<class T>
+Vector<T>::Vector()
+{
+	init();
+}
+
+template<class T>
+Vector<T>::Vector(SGVector<T> const &vector)
+{
+	init();
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector + vector.vlen, m_data.get());
+	m_len = vector.vlen;
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T> const &vector)
+{
+	init();
+
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+									[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+}
+
+template<class T>
+Vector<T>::~Vector()
+{
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data.get());
+	m_len = vector.vlen;
+	m_onGPU = false;
+	m_gpu_impl.release();
+	return *this;
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(Vector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+	 							[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl.reset(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		m_onGPU = false;
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+	return *this;
+}
+
+template<class T>
+Vector<T>::operator SGVector<T>()
+{
+	if (onGPU())
+		transferToCPU();
+	SGVector<T> vector(m_len);
+	std::copy(m_data.get(), m_data.get()+m_len, vector.vector);
+	return vector;
+}
+
+template<class T>
+bool Vector<T>::onGPU() const
+{
+	return m_onGPU;
+}
+
+template<class T>
+T* Vector<T>::data()
+{
+	if (onGPU())
+		transferToCPU();
+	return m_data.get();
+}
+
+template<class T>
+T const * Vector<T>::data() const
+{
+	return m_data.get();
```

Comment:
I have a copy constructor `Vector<T>::Vector(Vector<T> const &vector)`, which requires this const method.

However, the CPU and GPU cannot be synchronized within const method.

This might not be a big problem, because if the data in onGPU, the new Vector will have synced GPU and CPU vector. But the old vector won't and they'll be different. Also the const and non-const methods are incoherent.